File: Algorithms/Python3.x/199-Binary_Tree_Right_Side_View.py
# Runtime: 36 ms, faster than 82.58% of Python3 online submissions for Binary Tree Right Side View.
# Memory Usage: 13.6 MB, less than 5.26% of Python3 online submissions for Binary Tree Right Side View.
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        if root is None:
            return []
        self.res = []
        self.bfs([root])
        return self.res

    def bfs(self, layer):
        if layer == []:
            return
        new_layer = []
        flag = False  # is add node.val?
        for node in layer:
            if node and not flag:
                self.res.append(node.val)
                flag = True
            if node.right:
                new_layer.append(node.right)
            if node.left:
                new_layer.append(node.left)
        self.bfs(new_layer)

File: Algorithms/C++/199-Binary_Tree_Right_Side_View.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        std::vector<TreeNode*> layer = { root };
        std::vector<int> result;
        while (!layer.empty()) {
            std::vector<TreeNode*> nextLayer;
            bool foundEndNode = false;
            for (int i = layer.size() - 1; i >= 0; --i) {
                if (layer[i] != nullptr) {
                    if (!foundEndNode) {
                        result.push_back(layer[i]->val);
                        foundEndNode = true;
                    }
                    nextLayer.push_back(layer[i]->right);
                    nextLayer.push_back(layer[i]->left);
                }
            }
            layer = nextLayer;
            std::reverse(layer.begin(), layer.end());
        }
        return result;
    }
};
